// ng-auto-complete/esm2015/lib/classes/AutocompleteItem.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function StrippedAutocompleteGroup() { }
if (false) {
    /** @type {?} */
    StrippedAutocompleteGroup.prototype.group;
    /** @type {?} */
    StrippedAutocompleteGroup.prototype.item;
}
export class AutocompleteItem {
    /**
     * @param {?} title
     * @param {?} id
     * @param {?} className
     * @param {?} original
     */
    constructor(title, id, className, original) {
        this.title = title;
        this.id = id;
        this.className = className;
        this.original = original;
    }
}
if (false) {
    /** @type {?} */
    AutocompleteItem.prototype.title;
    /** @type {?} */
    AutocompleteItem.prototype.id;
    /** @type {?} */
    AutocompleteItem.prototype.children;
    /** @type {?} */
    AutocompleteItem.prototype.original;
    /** @type {?} */
    AutocompleteItem.prototype.className;
}
/**
 *
 * @param {?} items
 * @param {?} titleKey
 * @param {?=} childrenKey
 * @return {?}
 */
export function SearchableAutoCompleteItems(items, titleKey, childrenKey = null) {
    return items.reduce(function (r, i) {
        /** @type {?} */
        const t = SearchableAutoCompleteString(i[titleKey], i.id);
        if (typeof r[t] === 'undefined') {
            r[t] = TransformToAutocompleteItem(i, titleKey, childrenKey);
        }
        return r;
    }, {});
}
/**
 *
 * @param {?} key
 * @param {?} id
 * @return {?}
 */
export function SearchableAutoCompleteString(key, id) {
    return `${key.replace(/ /g, '_')}_id_${String(id)}`;
}
/**
 *
 * @param {?} str
 * @return {?}
 */
export function ComparableAutoCompleteString(str) {
    return str.replace(/_/g, ' ');
}
/**
 * object must have an ID
 * @param {?} object
 * @param {?} titleKey
 * @param {?=} childrenKey
 * @return {?}
 */
export function TransformToAutocompleteItem(object, titleKey, childrenKey = null) {
    /** @type {?} */
    const item = new AutocompleteItem(object[titleKey], object.id, object.className, object);
    /**
     * if there are children, add these.
     */
    if (childrenKey !== null) {
        item.children = object[childrenKey];
    }
    return item;
}
